cuda_core/tests/test_utils.py

```diff
@@ -0,0 +1,127 @@
+
+try:
+    import cupy as cp
+except ImportError:
+    cp = None
+try:
+    from numba import cuda as numba_cuda
+except ImportError:
+    numba_cuda = None
+import numpy as np
+import pytest
+
+from cuda.core.experimental import Device
+from cuda.core.experimental.utils import StridedMemoryView, args_viewable_as_strided_memory
+
+
+def convert_strides_to_counts(strides, itemsize):
+    return tuple(s // itemsize for s in strides)
+
+
+@pytest.mark.parametrize(
+    "in_arr,",
+    (
+        np.empty(3, dtype=np.int32),
+        np.empty((6, 6), dtype=np.float64)[::2, ::2],
+        np.empty((3, 4), order="F"),
+    ),
+)
+class TestViewCPU:
+    def test_args_viewable_as_strided_memory_cpu(self, in_arr):
+        @args_viewable_as_strided_memory((0,))
+        def my_func(arr):
+            # stream_ptr=-1 means "the consumer does not care"
+            view = arr.view(-1)
+            self._check_view(view, in_arr)
+
+        my_func(in_arr)
+
+    def test_strided_memory_view_cpu(self, in_arr):
+        # stream_ptr=-1 means "the consumer does not care"
+        view = StridedMemoryView(in_arr, stream_ptr=-1)
+        self._check_view(view, in_arr)
+
+    def _check_view(self, view, in_arr):
+        assert isinstance(view, StridedMemoryView)
+        assert view.ptr == in_arr.ctypes.data
+        assert view.shape == in_arr.shape
+        strides_in_counts = convert_strides_to_counts(in_arr.strides, in_arr.dtype.itemsize)
+        if in_arr.flags.c_contiguous:
+            assert view.strides is None
+        else:
+            assert view.strides == strides_in_counts
+        assert view.dtype == in_arr.dtype
+        assert view.device_id == -1
+        assert view.is_device_accessible is False
+        assert view.exporting_obj is in_arr
+
+
+def gpu_array_samples():
+    # TODO: this function would initialize the device at test collection time
+    samples = []
+    if cp is not None:
+        samples += [
+            (cp.empty(3, dtype=cp.complex64), None),
+            (cp.empty((6, 6), dtype=cp.float64)[::2, ::2], True),
+            (cp.empty((3, 4), order="F"), True),
+        ]
+    # Numba's device_array is the only known array container that does not
+    # support DLPack (so that we get to test the CAI coverage).
+    if numba_cuda is not None:
+        samples += [
+            (numba_cuda.device_array((2,), dtype=np.int8), None),
+            (numba_cuda.device_array((4, 2), dtype=np.float32), True),
+        ]
+    return samples
+
+
+def gpu_array_ptr(arr):
+    if cp is not None and isinstance(arr, cp.ndarray):
+        return arr.data.ptr
+    if numba_cuda is not None and isinstance(arr, numba_cuda.cudadrv.devicearray.DeviceNDArray):
+        return arr.device_ctypes_pointer.value
+    raise NotImplementedError(f"{arr=}")
+
+
+@pytest.mark.parametrize("in_arr,stream", (*gpu_array_samples(),))
+class TestViewGPU:
+    def test_args_viewable_as_strided_memory_gpu(self, in_arr, stream):
```

Comment:
Instead of `stream` maybe call it `use_stream`?  And then having the gpu-array_samples return False instead of None. Same change for other tests here.

I got confused for a little since the naming implied to me that it's a stream handle.

Reply:
Good idea, done in commit 16fc9f69.